### ISSUE TITLE
Tooltip on Buttons

### ISSUE THREAD

**Issue by Jigisuta** [beeware/toga#2598]
### What is the problem or limitation you are having?

Hi Team, 

I was searching the internet for any reference to tooltips on buttons, but I couldn't find any. For context, whenever we use only icons (no text) for buttons, it may be possible that the meaning of that icon may not be widely known, then in that case having a tooltip will greatly help the user of the app to understand what that icon mean and what could be the functionality of that icon. 

My Icon:

![image](https://github.com/beeware/toga/assets/103924256/6ab7f512-ffc6-43a4-ad41-3a94bb68535f)

This feature is available in toga.command, but that one is also not working for me, below is my code:

`connection = toga.Group("Connections")

editCon = toga.Command(
    editOracleConAction,
    text="Edit Connection",
    tooltip="Create, Edit, Test & Save Connections",
    group=connection,
    )   
self.commands.add(editCon)`

![image](https://github.com/beeware/toga/assets/103924256/10e0d1a8-eb63-4d01-af94-06decd5859a2)

**If there is some solution to this, please point me in the correct direction.** 

### Describe the solution you'd like

Whenever the user hovers over the icon, it should display the tooltip information. 

### Describe alternatives you've considered

An alternative is to use text along with icons, but that is not what I am looking for. Using text will defeat the purpose of using an icon.

**Comment by freakboy3742** [comment]
Agreed that this could be useful, especially in the icon button case. The tooltip option on command is only used where a tooltip provides additional help - in particular, the use of commands on toolbars, where they are often implemented as icon buttons.

This could actually pair well with the [need to add accessibility hints](https://github.com/beeware/toga/issues/2233#issuecomment-1817698460) for text inputs; the label used for accessibility could easily serve double-purpose as a tooltip.

**Comment by jgilmore** [comment]
Seconded. I, also, searched for this functionality (which is common in desktop GUI toolkits) and found there's just no way at all to do this currently.

I know that GTK, at least, offers this function on pretty much ALL widgets. It probably makes sense to attach this to the widget class rather than adding it piecemeal where supported by particular widgets.

**Comment by ajkessel** [comment]
I'm considering porting my application from customtkinter to Toga but this is a blocker. Whether the button is a short word or an icon, I would like to give the user hints about what the button does (including advertising the shortcut key for the button).

Is there a workaround that would accomplish this with the current Toga features?

**Comment by freakboy3742** [comment]
@ajkessel Any workaround at present would be based on calling platform-specific APIs. Any Toga widget retains a reference to the underlying native widget; and you can invoke native APIs on that native widget. 

So - on macOS, you could use:
```
button = toga.Button("My Button")
button._impl.native.toolTip = "My button tooltip"
```

and the button will gain a tooltip with the given text. If you're only targeting macOS, that's all you need to do; if you're planning on your app being multi-platform, you'll need to accommodate that. I haven't tested the GTK and Windows code, but it should look something like:
```
button = toga.Button("My Button")
if sys.platform == "darwin":
    button._impl.native.toolTip = "My button tooltip"
elif sys.platform == "linux":
    button._impl.native.set_tooltip_text("My button tooltip")
elif sys.platform == "win32":
    import System.Windows.Forms import ToolTip
    tooltip = Tooltip()
    tooltip.SetToolTip(button._impl.native, "My button tooltip")
```

**Comment by ajkessel** [comment]
> If you're only targeting macOS, that's all you need to do; if you're planning on your app being multi-platform, you'll need to accommodate that. I haven't tested the GTK and Windows code, but it should look something like:

Not too bad! Probablybetter than all the manual tweaks I've implemented trying to get customtkinter tooltips to look right across platforms.

**Comment by alej0varas** [comment]
Hi! I've implemented this for a project and it works. Ready to copy/paste code below ;). Tested only in Linux with GTK for `Switch` and `Button` widgets.
`set_tooltip` uses @freakboy3742's [code](https://github.com/beeware/toga/issues/2598#issuecomment-4443187518). I've placed the code in my `app.py` file where `toga` is imported for the first time.

```
# Keep your import where it is.
import toga

# Put the rest of the code before instantiating any toga widget.
class MonkeyPatchedWidget(toga.Widget):
    def __init__(self, *args, tooltip=None, **kwargs):
        super().__init__(*args, **kwargs)

        if tooltip:
            self.set_tooltip(tooltip)

    def set_tooltip(self, tooltip):
        import sys
        if sys.platform == "darwin":
            self._impl.native.toolTip = tooltip
        elif sys.platform == "linux":
            self._impl.native.set_tooltip_text(tooltip)
        elif sys.platform == "win32":
            from System.Windows.Forms import ToolTip
            tooltip = ToolTip()
            tooltip.SetToolTip(self._impl.native, tooltip)


toga.widgets.base.Widget = MonkeyPatchedWidget
```

Usage
```
button = toga.Button('my button', on_press=on_press_handler, tooltip='Some helpful text')
```

Please let me know if this can be added to toga so I'll work on a PR. 
And if anybody knows what is the equivalent for mobile implementations please comment how to do it so I can include it.

**Comment by freakboy3742** [comment]
The ticket is open, and I've indicated in the discussion that the idea has merit - which means a PR with a good implementation would be accepted. That implementation will use many of the same *pieces* as the monkey patch, but it will be distributed across the backends, and there is more needed - including tests and documentation.
